src/resolvers/Mutation.js

```diff
@@ -0,0 +1,157 @@
+
+async function signup(parent, args, context) {
+  const password = await bcrypt.hash(args.password, 10);
+  const user = await context.prisma.createUser({ ...args, password });
+  const token = jwt.sign({ userId: user.id }, APP_SECRET);
+
+  return {
+    token,
+    user
+  };
+}
+
+async function login(parent, args, context) {
+  const user = await context.prisma.user({ email: args.email });
+  if (!user) {
+    throw new Error("No such user found");
+  }
+
+  const valid = await bcrypt.compare(args.password, user.password);
+  if (!valid) {
+    throw new Error("Invalid password");
+  }
+
+  return {
+    token: jwt.sign({ userId: user.id }, APP_SECRET),
+    user
+  };
+}
+
+function postBudget(parent, args, context) {
+  const userId = getUserId(context);
+  return context.prisma.createBudget({
+    total: args.total,
+    startDate: args.startDate,
+    endDate: args.endDate,
+    savingsTarget: args.savingsTarget,
+    postedBy: { connect: { id: userId } }
+  });
+}
+
+async function postExpense(parent, args, context) {
+  const userId = getUserId(context);
+  const expense = await context.prisma.createExpense({
+    expenseAmount: args.expenseAmount,
+    expenseDesc: args.expenseDesc,
+    expenseCategory: args.expenseCategory,
+    user: { connect: { id: userId } },
+    budget: { connect: { id: args.budgetId } }
+  });
+  const budget = await context.prisma.budget({
+    id: args.budgetId
+  });
+  const allExpenses = await context.prisma.expenses({
+    where: {
+      user: {
+        id: userId
+      },
+      budget: {
+        id: args.budgetId
+      }
+    }
+  });
+  const totalExpenses = allExpenses.reduce((acc, curr) => {
+    if (curr.expenseAmount) {
+      return (acc = acc + curr.expenseAmount);
+    }
+  }, 0.0);
+  const remainingBudget = budget.total - totalExpenses;
+  context.prisma.updateBudget({
+    data: {
+      remainingAmount: remainingBudget,
+      totalExpenses: totalExpenses
+    },
+    where: {
+      id: args.budgetId
+    }
+  });
+  return expense;
+}
+
+async function postSavings(parent, args, context) {
+  const userId = getUserId(context);
+  const matchingBudget = await context.prisma.budget({
+    id: args.budgetId
+  });
+  const matchingBudgetSavings = await context.prisma
+    .budget({
+      id: args.budgetId
+    })
+    .savings();
+  let savingsRecord = {};
+  if (matchingBudgetSavings.amount === undefined) {
+    savingsRecord = await context.prisma.createSavings({
+      amount: args.amount,
+      user: { connect: { id: userId } },
+      budget: { connect: { id: args.budgetId } }
+    });
+  } else {
+    savingsRecord = await context.prisma.updateSavings({
+      data: {
+        amount: args.amount
+      },
+      where: {
+        id: matchingBudgetSavings.id
+      }
+    });
+  }
+  const budgets = await context.prisma.budgets({
+    where: {
+      postedBy: {
+        id: userId
+      }
+    },
+    orderBy: "endDate_DESC"
+  });
+  budgets.map(async budget => {
```

Comment:
this is a nice one. Whenever we use array methods we prefer that they don't have "side-effects". These can be for example making calls to the DB for creation of stuff. Map method is to transform one array into another and we should keep it that way so you have two options:
- You keep using map but you map the array into an array of promises and then use Promise.all to await on them all, which makes it evident that the new array is going to perform other stuff outside of just transforming one value into another.
- You switch to using `.forEach` which reads better for this as "for each of these do this"

Reply:
this is a nice reference: https://medium.com/@ntgard/foreach-is-for-side-effects-60fca9f78850

Reply:
I will start using for each